### PR TITLE
Issue 174

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

**Comment by je-ik** [comment]
This closes O2-Czech-Republic#174

File: core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java [hunk 1]
```diff
@@ -122,6 +123,7 @@ public void close() {
     if (tracker != null) {
       ExceptionUtils.unchecked(tracker::close);
     }
+    lastGlobalWatermarkUpdate = Long.MIN_VALUE;
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java:126]: what is this good for? we don't reuse instances right?

File: core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java [hunk 2]
```diff
@@ -138,16 +140,18 @@ public String toString() {
...
       lastGlobalWatermarkUpdate = now;
-      log.debug("Updated watermark of {} to {}", this, globalWatermark);
     }
-    globalWatermark = tracker.getGlobalWatermark(processName, context.getMinWatermark());
   }
 
   @VisibleForTesting
   void forceUpdateGlobalWatermark() {
-    lastGlobalWatermarkUpdate = 0;
+    lastGlobalWatermarkUpdate = Long.MIN_VALUE;
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java:154]: This seems like a workaround for not using `Clock` instead of `System.currentTimeMillis()` ...

File: direct/io-hbase-bindings/src/main/java/cz/o2/proxima/direct/hbase/HBaseLogReader.java [hunk 3]
```diff
@@ -119,32 +136,29 @@ public void observe(
   private void flushPartitions(
       List<Partition> partitions,
       List<AttributeDescriptor<?>> attributes,
+      TerminationContext terminationContext,
       BatchLogObserver observer)
       throws IOException {
 
+    outer:
```
Comment (dmvk) [inline direct/io-hbase-bindings/src/main/java/cz/o2/proxima/direct/hbase/HBaseLogReader.java:143]: 👎

File: direct/core/src/test/resources/log4j.properties [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Root logger option
-log4j.rootLogger=INFO, stdout
-#log4j.rootLogger=DEBUG, stdout
```
Comment (dmvk) [inline direct/core/src/test/resources/log4j.properties:3]: these changes don't belong in PRs, this is changed in literally every 2nd PR ...

File: direct/core/src/main/java/cz/o2/proxima/direct/commitlog/CommitLogReaders.java [hunk 5]
```diff
@@ -129,14 +130,12 @@ public boolean hasExternalizableOffsets() {
 
     @Getter private final ThroughputLimiter limiter;
     private final Collection<Partition> partitions;
-    private final int numPartitions;
     private long watermark = Long.MIN_VALUE;
 
     public LimitedCommitLogReader(CommitLogReader delegate, ThroughputLimiter limiter) {
       super(delegate);
-      this.limiter = Objects.requireNonNull(limiter);
+      this.limiter = SerializableUtils.clone(Objects.requireNonNull(limiter));
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/commitlog/CommitLogReaders.java:137]: why is this necessary? either way, this should be caller's responsibility to provide healthy limiter instance

File: direct/core/src/main/java/cz/o2/proxima/direct/commitlog/CommitLogReaders.java [hunk 6]
```diff
@@ -198,21 +197,29 @@ private LogObserver throughputLimited(LogObserver delegate) {
 
         @Override
         public void onCompleted() {
-          super.onCompleted();
-          limiter.close();
+          try {
+            super.onCompleted();
+          } finally {
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/commitlog/CommitLogReaders.java:202]: 👍

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java [hunk 7]
```diff
@@ -115,34 +115,45 @@ private ForwardingBatchLogObserver(BatchLogObserver delegate) {
...
         BatchLogObserver delegate,
-        int numPartitions,
         Collection<Partition> assignedPartitions,
         ThroughputLimiter limiter) {
 
       super(delegate);
-      this.numPartitions = numPartitions;
       this.assignedPartitions = new ArrayList<>(assignedPartitions);
-      this.limiter = Objects.requireNonNull(limiter);
+      this.limiter = SerializableUtils.clone(Objects.requireNonNull(limiter));
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java:129]: same as in commit log observer

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReader.java [hunk 8]
```diff
@@ -66,7 +66,7 @@
    * @param attributes attribute descriptors to filter out
    * @param observer the observer by which to consume the data
    */
-  void observe(
+  ObserveHandle observe(
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReader.java:69]: missing `@return` javadoc

File: beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java [hunk 9]
```diff
@@ -1118,7 +1118,7 @@ private static void registerCommonTypes(Kryo kryo, Repository repo) {
     for (Field f : obj.getClass().getDeclaredFields()) {
```
Comment (dmvk) [inline beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java:1118]: this code seems weird... the most used serializer is protobuf, which should be registered as:

```
kryo.register(message, new ProtobufSerializer<>)
```

current code results into not using proto serialization when available

File: beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java [hunk 10]
```diff
@@ -1411,10 +1425,18 @@ public void onTimer(
     }
   }
 
-  private static <T> DoFn<T, Void> asDoFn(Consumer<T> consumer) {
+  private <T> DoFn<T, Void> asDoFn(PipelineOptions opts, Consumer<T> consumer) {
+    if (supportsStableInput(opts)) {
```
Comment (dmvk) [inline beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java:1429]: why is this needed? this would result in increased latency when consuming remote output

File: core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java (hunk 1 above)
Comment (je-ik) [reply core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java:126]: We were reusing them, but we are cloning them instead now. The reuse was not thread-safe. This is a left-over.

File: core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java (hunk 2 above)
Comment (je-ik) [reply core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java:154]: I would not use `Clock`. We could create a `TimeProvider` in core, but that seems like overkill for what we need here.

File: direct/core/src/main/java/cz/o2/proxima/direct/commitlog/CommitLogReaders.java (hunk 5 above)
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/commitlog/CommitLogReaders.java:137]: Hmm... This looks strange. My intent definitely was to clone this into the `LogObserver`. This looks like a bug.

File: direct/io-hbase-bindings/src/main/java/cz/o2/proxima/direct/hbase/HBaseLogReader.java (hunk 3 above)
Comment (je-ik) [reply direct/io-hbase-bindings/src/main/java/cz/o2/proxima/direct/hbase/HBaseLogReader.java:143]: Simplyfying code, nothing wrong with this.

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java (hunk 7 above)
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java:129]: yes

File: beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java (hunk 9 above)
Comment (je-ik) [reply beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java:1118]: unrelated to this PR.

Comment (je-ik) [reply beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java:1118]: Most notably - protobuf is registered as Coder using `BeamStream#registerCodersForSchemes`. This just registers additional types in Kryo.

File: beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java (hunk 10 above)
Comment (je-ik) [reply beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java:1429]: Because otherwise the consumer will see duplicates.

File: core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java (hunk 1 above)
Comment (je-ik) [reply core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkThroughputLimiter.java:126]: Fixed.

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReader.java (hunk 8 above)
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReader.java:69]: Fixed.

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java (hunk 7 above)
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java:129]: This is actually correct. The problem is that `LogReader` (which owns the `ThroughputLimiter`) has different life-cycle than the `LogObserver`. You can have multiple observers observe the same reader and therefore each needs its own instance of limiter.

File: direct/core/src/main/java/cz/o2/proxima/direct/commitlog/CommitLogReaders.java (hunk 5 above)
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/commitlog/CommitLogReaders.java:136]: Okay, definitely a bug. The clone should have been in different place. Fixed.

File: direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java [hunk 11]
```diff
@@ -66,13 +69,14 @@
...
   private static final AtomicInteger threadId = new AtomicInteger();
+  private static final String ID = UUID.randomUUID().toString();
 
   private static Factory<ExecutorService> createExecutorFactory() {
     return () ->
         Executors.newCachedThreadPool(
             r -> {
               Thread t = new Thread(r);
-              t.setName(String.format("DirectDataOperatorThread-%d", threadId.incrementAndGet()));
+              t.setName(String.format("DirectDataOperatorThread-%s-%d", ID, threadId.incrementAndGet()));
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java:79]: what is purpose of this? the ID is static, so it will remain the same for all instances within JVM

File: direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java [hunk 12]
```diff
@@ -50,8 +50,11 @@
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.UUID;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.SynchronousQueue;
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java:56]: unused? how comes spotless did not delete these?

File: direct/io-blob/src/main/java/cz/o2/proxima/direct/blob/BlobLogReader.java [hunk 13]
```diff
@@ -199,66 +205,95 @@ protected BlobLogReader(BlobStorageAccessor accessor, Context context) {
...
+            Iterator<Partition> iterator = partitions.iterator();
+            while (iterator.hasNext()
+                && !terminationContext.isCancelled()
+                && !stopProcessing.get()) {
+              Partition p = iterator.next();
+              processSinglePartition(p, attrs, terminationContext, stopProcessing, observer);
             }
+            terminationContext.finished();
+          } catch (Throwable ex) {
+            terminationContext.handleErrorCaught(
```
Comment (dmvk) [inline direct/io-blob/src/main/java/cz/o2/proxima/direct/blob/BlobLogReader.java:243]: `onError` code path doesn't seem covered in `BlobLogReaderTest`

File: direct/io-cassandra/src/main/java/cz/o2/proxima/direct/cassandra/CassandraLogReader.java [hunk 14]
```diff
@@ -69,68 +71,91 @@
...
+    for (Row row : result) {
+      if (terminationContext.isCancelled()) {
+        return false;
+      }
+      String key = row.getString(0);
+      int field = 1;
+      for (AttributeDescriptor<?> attribute : attributes) {
+        String attributeName = attribute.getName();
+        if (attribute.isWildcard()) {
+          // FIXME: this is wrong
```
Comment (dmvk) [inline direct/io-cassandra/src/main/java/cz/o2/proxima/direct/cassandra/CassandraLogReader.java:128]: do we have an issue for this?

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/TerminationContext.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+import cz.o2.proxima.util.ExceptionUtils;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+
+/** A context that guards correct termination of batch observation process. */
+@Internal
+@Slf4j
+public class TerminationContext {
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/batch/TerminationContext.java:28]: 👍 for this context

File: direct/core/src/main/java/cz/o2/proxima/direct/batch/TerminationContext.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+  private final CountDownLatch terminateLatch = new CountDownLatch(1);
+  private volatile boolean cancelled = false;
+  private volatile Thread runningThread;
+
+  public TerminationContext(BatchLogObserver observer) {
+    this.observer = observer;
+  }
+
+  public boolean isCancelled() {
+    return cancelled || runningThread == Thread.currentThread() && runningThread.isInterrupted();
```
Comment (dmvk) [inline direct/core/src/main/java/cz/o2/proxima/direct/batch/TerminationContext.java:40]: can you elaborate little bit about why it not enough to check only `cancelled` flag here?

Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/batch/TerminationContext.java:40]: Some runners (e.g. flink) might interrupt the running thread to terminate it.

File: direct/io-cassandra/src/main/java/cz/o2/proxima/direct/cassandra/CassandraLogReader.java (hunk 14 above)
Comment (je-ik) [reply direct/io-cassandra/src/main/java/cz/o2/proxima/direct/cassandra/CassandraLogReader.java:128]: I'd say we do not. This is not modified in this PR.

File: direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java (hunk 11 above)
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java]: It is the same for all instances of the same classloader.

File: direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java (hunk 12 above)
Comment (je-ik) [reply direct/core/src/main/java/cz/o2/proxima/direct/core/DirectDataOperator.java]: Probably before spotless was applied.

**Review by dmvk — APPROVED** [review]
LGTM 👍 Next time we should focus on keeping the change set small as this is impossible to properly review. Good job ;)
